File: lib/loka/core/compose.ex
defmodule Loka.Core.Compose do
  @moduledoc """
  StateDelta composition (04 §5.1-§5.4, 14 §R3A). `kernel/ts/src/compose.ts` is the
  TypeScript twin; both run `protocol/fixtures/composition.json`.

  Takes a committed base state and a contract-valid StateDelta. Ops apply in their semantic
  order to a proposal overlay; each op's precondition reads the overlay over the base. An op
  on a target another writer group already wrote faults `conflicting_write` (no
  last-writer-wins, no same-value coalescing). The base is never copied.

  Base state, a JSON object (absent sections are empty):

  - `"clock"`: the committed logical time;
  - `"facts"`: canonical fact MutationTarget text => FactValue (compared as opaque values);
  - `"fact_defaults"`: canonical DefinitionRef text => FactValue for unset facts;
  - `"containers"`: EntityId => its one container (03 §23); `"capacities"`: EntityId =>
    the most entities it may contain;
  - `"quests"`: QuestInstanceId => `quest`, `scope`, `state`, optional `outcome`;
  - `"choices"`: ContinuationId => the `choice.open` fields plus `status` and the
    host-assigned `opened_revision`;
  - `"jobs"`: JobId => `job`, `due_time`, `status`.

  Result: `%{"changes" => rows}`, one `%{"target", "value"}` per written MutationTarget,
  sorted by canonical target text (the rows the host commits, ADR-072; a new continuation
  lacks `opened_revision` until the host commits it), or `%{"fault" => fault}`, a fault
  DecisionResult naming the op's target when there is one. Budgets from the composition
  profile are checked first, then ops in order: conflict before precondition.

  An explicit advance is a delta with `time.advance`; its target (the last `to`) is the
  visited time for `job.complete` and the bound `job.schedule` must exceed (04 §5.4).
  Without one, both use the base clock.
  """
  alias Loka.Core.Canonical

  @profile_path Path.expand("../../../docs/spec/conformance/composition-profile.json", __DIR__)
  @external_resource @profile_path
  @profile File.read!(@profile_path)
  @limits JSON.decode!(@profile)["limits"]

  @legal %{
    "active" => ~w(objectives_complete failed abandoned),
    "objectives_complete" => ~w(resolved failed abandoned),
    "failed" => ["active"],
    "abandoned" => ["active"]
  }

  @spec compose(map(), map()) :: %{String.t() => term()}
  def compose(state, %{"ops" => ops}) do
    if over_budget?(state, ops) do
      fault("budget_exceeded", nil)
    else
      horizon = Enum.reduce(ops, state["clock"], &advance_target/2)

      case Enum.reduce_while(ops, %{}, &step(&1, &2, {state, horizon})) do
        %{"fault" => _} = f -> f
        overlay -> %{"changes" => overlay |> Enum.sort() |> Enum.map(&row/1)}
      end
    end
  end

  @doc "The MutationTarget an op writes (04 §5.1)."
  @spec target(map()) :: map()
  def target(%{"op" => "fact.assign"} = op),
    do: Map.put(Map.take(op, ~w(fact scope subject_id)), "kind", "fact")

  def target(%{"op" => "entity.transfer", "entity_id" => e}), do: containment(e)

  def target(%{"op" => "quest." <> _, "instance_id" => i}),
    do: %{"kind" => "quest", "instance_id" => i}

  def target(%{"op" => "choice." <> _, "continuation_id" => c}),
    do: %{"kind" => "choice", "continuation_id" => c}

  def target(%{"op" => "job." <> _, "job_id" => j}), do: %{"kind" => "job", "job_id" => j}
  def target(%{"op" => "time.advance"}), do: %{"kind" => "clock"}

  @doc "Canonical text of a JSON value: the identity of a target or DefinitionRef."
  @spec key(term()) :: binary()
  def key(value) do
    {:ok, text} = Canonical.encode(value)
    text
  end

  defp over_budget?(state, ops) do
    count = fn name -> Enum.count(ops, &(&1["op"] == name)) end
    pending = Enum.count(section(state, "jobs"), fn {_, j} -> j["status"] == "pending" end)
    created = count.("job.schedule")
    due = count.("job.complete")

    length(ops) > @limits["operations"] or created > @limits["created_jobs"] or
      due > @limits["due_jobs_per_advance"] or pending + created - due > @limits["pending_jobs"]
  end

  defp advance_target(%{"op" => "time.advance", "to" => to}, _), do: to
  defp advance_target(_, t), do: t

  defp step(op, overlay, {state, horizon}) do
    target = target(op)
    k = key(target)
    group = op["writer_group"]

    case overlay do
      %{^k => {other, _, _}} when other != group -> {:halt, fault("conflicting_write", target)}
      _ -> apply_op(op, target, {state, horizon, overlay}) |> write(overlay, k, group, target)
    end
  end

  defp write({:ok, value}, overlay, k, group, target),
    do: {:cont, Map.put(overlay, k, {group, target, value})}

  defp write({:error, code}, _, _, _, target), do: {:halt, fault(code, target)}

  defp apply_op(%{"op" => "fact.assign"} = op, t, ctx) do
    now = with nil <- read(t, ctx), do: get_in(elem(ctx, 0), ["fact_defaults", key(op["fact"])])
    check(now == op["expected"], op["value"])
  end

  defp apply_op(
         %{"op" => "entity.transfer", "entity_id" => e, "destination_id" => d} = op,
         t,
         ctx
       ) do
    cond do
      read(t, ctx) != op["source_id"] ->
        {:error, "precondition_failed"}

      inside?(d, e, ctx, map_size(section(elem(ctx, 0), "containers")) + 1) ->
        {:error, "containment_cycle"}

      full?(d, e, ctx) ->
        {:error, "capacity_exceeded"}

      true ->
        {:ok, d}
    end
  end

  defp apply_op(%{"op" => "quest.activate", "quest" => q, "scope" => s}, t, ctx) do
    open? = fn {_, r} ->
      r["quest"] == q and r["scope"] == s and r["state"] in ~w(active objectives_complete)
    end

    taken = Enum.any?(rows("quest", "quests", "instance_id", ctx), open?)
    check(read(t, ctx) == nil and not taken, %{"quest" => q, "scope" => s, "state" => "active"})
  end

  defp apply_op(%{"op" => "quest.transition", "from" => from, "to" => to} = op, t, ctx) do
    row = read(t, ctx)
    outcome = op["outcome"]
    outcome_ok = if to == "resolved", do: outcome != nil, else: to == "failed" or outcome == nil

    next =
      if outcome,
        do: Map.put(row || %{}, "outcome", outcome),
        else: Map.delete(row || %{}, "outcome")

    check(
      row["state"] == from and to in Map.get(@legal, from, []) and outcome_ok,
      Map.put(next, "state", to)
    )
  end

  defp apply_op(%{"op" => "choice.open"} = op, t, ctx) do
    row = Map.take(op, ~w(actor_id source beat roles choice_ids))
    check(read(t, ctx) == nil, Map.put(row, "status", "pending"))
  end

  defp apply_op(%{"op" => "choice.resolve", "choice_id" => c} = op, t, ctx) do
    row = read(t, ctx)

    check(
      row["status"] == "pending" and c in row["choice_ids"] and
        row["opened_revision"] == op["expected_revision"],
      Map.merge(row || %{}, %{"status" => "resolved", "choice_id" => c})
    )
  end

  defp apply_op(%{"op" => "choice.close"}, t, ctx) do
    row = read(t, ctx)
    check(row["status"] == "pending", Map.put(row || %{}, "status", "closed"))
  end

  defp apply_op(%{"op" => "job.schedule", "due_time" => due} = op, t, {_, horizon, _} = ctx) do
    cond do
      read(t, ctx) != nil -> {:error, "precondition_failed"}
      due <= horizon -> {:error, "nonfuture_job"}
      true -> {:ok, %{"job" => op["job"], "due_time" => due, "status" => "pending"}}
    end
  end

  defp apply_op(%{"op" => "job.complete"}, t, {_, horizon, _} = ctx) do
    row = read(t, ctx)

    check(
      row["status"] == "pending" and row["due_time"] <= horizon,
      Map.put(row || %{}, "status", "completed")
    )
  end

  defp apply_op(%{"op" => "time.advance", "from" => from, "to" => to}, t, ctx),
    do: check(read(t, ctx) == from and to > from, to)

  defp check(true, value), do: {:ok, value}
  defp check(false, _), do: {:error, "precondition_failed"}

  # The overlay's value for a target, else the base's.
  defp read(t, {state, _, overlay}) do
    case Map.fetch(overlay, key(t)) do
      {:ok, {_, _, value}} -> value
      :error -> base(t, state)
    end
  end

  defp base(%{"kind" => "fact"} = t, s), do: section(s, "facts")[key(t)]
  defp base(%{"kind" => "containment", "entity_id" => e}, s), do: section(s, "containers")[e]
  defp base(%{"kind" => "quest", "instance_id" => i}, s), do: section(s, "quests")[i]
  defp base(%{"kind" => "choice", "continuation_id" => c}, s), do: section(s, "choices")[c]
  defp base(%{"kind" => "job", "job_id" => j}, s), do: section(s, "jobs")[j]
  defp base(%{"kind" => "clock"}, s), do: s["clock"]

  # d is e or inside it. A walk longer than the containment rows means a cyclic base: fail closed.
  defp inside?(nil, _, _, _), do: false
  defp inside?(e, e, _, _), do: true
  defp inside?(_, _, _, 0), do: true
  defp inside?(d, e, ctx, n), do: inside?(read(containment(d), ctx), e, ctx, n - 1)

  defp full?(d, e, {state, _, _} = ctx) do
    case section(state, "capacities")[d] do
      nil ->
        false

      cap ->
        Enum.count(
          rows("containment", "containers", "entity_id", ctx),
          &(elem(&1, 1) == d and elem(&1, 0) != e)
        ) >= cap
    end
  end

  # ponytail: scans the whole section; add a contents/scope index when a cartridge has many rows.
  defp rows(kind, name, id, {state, _, overlay}) do
    changed = for {_, {_, %{"kind" => ^kind} = t, v}} <- overlay, into: %{}, do: {t[id], v}
    Map.merge(section(state, name), changed)
  end

  defp section(state, name), do: Map.get(state, name, %{})
  defp containment(e), do: %{"kind" => "containment", "entity_id" => e}
  defp row({_, {_, target, value}}), do: %{"target" => target, "value" => value}

  defp fault(code, nil), do: %{"fault" => %{"kind" => "fault", "code" => code}}

  defp fault(code, target),
    do: %{"fault" => %{"kind" => "fault", "code" => code, "target" => target}}
end

File: kernel/ts/src/compose.ts
// StateDelta composition (04 §5.1-§5.4, 14 §R3A), twin of lib/loka/core/compose.ex, whose
// moduledoc states the base-state shape and the semantics. The base is read, never copied:
// writes go to an overlay keyed by canonical target text.
import { encode, type Json } from './canonical.ts';
import {
  LIMITS,
  type DeltaOp,
  type ErrorCode,
  type MutationTarget,
  type StateDelta,
} from './contracts.gen.ts';

type Obj = { readonly [key: string]: Json };
export type State = { readonly clock: number } & { readonly [section: string]: Json };
export type Change = { target: MutationTarget; value: Json };
export type Fault = { kind: 'fault'; code: ErrorCode; target?: MutationTarget };
export type Result = { changes: Change[] } | { fault: Fault };

type Written = { group: number; target: MutationTarget; value: Json };
type Ctx = { state: State; horizon: number; overlay: Map<string, Written> };
type Outcome = { value: Json } | { code: ErrorCode };

const LEGAL: Record<string, string[]> = {
  active: ['objectives_complete', 'failed', 'abandoned'],
  objectives_complete: ['resolved', 'failed', 'abandoned'],
  failed: ['active'],
  abandoned: ['active'],
};
const OPEN = ['active', 'objectives_complete'];

export const key = (value: unknown): string => encode(value as Json);
export const same = (a: unknown, b: unknown): boolean => key(a ?? null) === key(b ?? null);
const get = (o: Json | undefined, k: string): Json | undefined =>
  o !== null && typeof o === 'object' && !Array.isArray(o) && Object.hasOwn(o, k)
    ? (o as Obj)[k]
    : undefined;
const section = (s: State, name: string): Obj => (get(s, name) ?? {}) as Obj;
const containment = (e: string): MutationTarget =>
  ({ kind: 'containment', entity_id: e }) as MutationTarget;

/** The MutationTarget an op writes (04 §5.1). */
export function target(op: DeltaOp): MutationTarget {
  switch (op.op) {
    case 'fact.assign': {
      const t: Record<string, unknown> = { kind: 'fact', fact: op.fact, scope: op.scope };
      if (op.subject_id !== undefined) t.subject_id = op.subject_id;
      return t as MutationTarget;
    }
    case 'entity.transfer':
      return containment(op.entity_id);
    case 'quest.activate':
    case 'quest.transition':
      return { kind: 'quest', instance_id: op.instance_id };
    case 'choice.open':
    case 'choice.resolve':
    case 'choice.close':
      return { kind: 'choice', continuation_id: op.continuation_id };
    case 'job.schedule':
    case 'job.complete':
      return { kind: 'job', job_id: op.job_id };
    case 'time.advance':
      return { kind: 'clock' };
    case 'resource.adjust':
      return { kind: 'resource', resource: op.resource, entity_id: op.entity_id };
    case 'cooldown.start':
      return { kind: 'cooldown', actor_id: op.actor_id, action: op.action };
  }
}

/** A resource's stored row: its value and the time it was stored (delta.schema.json). */
export type Stored = { readonly value: number; readonly at: number };
type Spec = {
  readonly maximum: number;
  readonly minimum: number;
  readonly start: number;
  readonly gain: number;
};

/**
 * A resource's current value at `now` (ResourceSpec regeneration): the stored value (start at
 * time 0 when unset) plus gain for each hour boundary crossed since it was stored, stopping at
 * maximum. A product past 2^53 is inexact but still above maximum, so the result is exact.
 */
export function current(row: Stored | undefined, spec: Spec, now: number): number {
  const { value, at } = row ?? { value: spec.start, at: 0 };
  const ticks = Math.floor(now / 3600) - Math.floor(at / 3600);
  return Math.min(spec.maximum, value + spec.gain * ticks);
}

export function compose(state: State, delta: StateDelta): Result {
  const { ops } = delta;
  if (typeof state.clock !== 'number') return fault('precondition_failed', { kind: 'clock' });
  if (overBudget(state, ops)) return { fault: { kind: 'fault', code: 'budget_exceeded' } };
  let horizon = state.clock;
  for (const op of ops) if (op.op === 'time.advance') horizon = op.to;
  const ctx: Ctx = { state, horizon, overlay: new Map() };
  for (const op of ops) {
    const t = target(op);
    const k = key(t);
    const prior = ctx.overlay.get(k);
    if (prior && prior.group !== op.writer_group) return fault('conflicting_write', t);
    const out = apply(op, t, ctx);
    if ('code' in out) return fault(out.code, t);
    ctx.overlay.set(k, { group: op.writer_group, target: t, value: out.value });
  }
  const rows = [...ctx.overlay].sort(([a], [b]) => (a < b ? -1 : 1));
  return { changes: rows.map(([, w]) => ({ target: w.target, value: w.value })) };
}

function overBudget(state: State, ops: readonly DeltaOp[]): boolean {
  const count = (name: string) => ops.filter((o) => o.op === name).length;
  const jobs = Object.values(section(state, 'jobs'));
  const pending = jobs.filter((j) => get(j, 'status') === 'pending').length;
  const created = count('job.schedule');
  const due = count('job.complete');
  return (
    ops.length > LIMITS.operations! ||
    created > LIMITS.created_jobs! ||
    due > LIMITS.due_jobs_per_advance! ||
    pending + created - due > LIMITS.pending_jobs!
  );
}

const fault = (code: ErrorCode, t: MutationTarget): Result => ({
  fault: { kind: 'fault', code, target: t },
});
const check = (ok: boolean, value: Json): Outcome =>
  ok ? { value } : { code: 'precondition_failed' };
const put = (row: Json | undefined, extra: Obj): Json => ({ ...((row ?? {}) as Obj), ...extra });

function apply(op: DeltaOp, t: MutationTarget, ctx: Ctx): Outcome {
  const row = read(t, ctx);
  switch (op.op) {
    case 'fact.assign': {
      const now = row ?? get(section(ctx.state, 'fact_defaults'), key(op.fact));
      return check(now !== undefined && same(now, op.expected), op.value);
    }
    case 'entity.transfer':
      return transfer(op.entity_id, op.source_id, op.destination_id, row, ctx);
    case 'quest.activate':
    case 'quest.transition':
      return quest(op, row, ctx);
    case 'choice.open':
    case 'choice.resolve':
    case 'choice.close':
      return choice(op, row);
    case 'job.schedule':
      if (row !== undefined) return { code: 'precondition_failed' };
      if (op.due_time <= ctx.horizon) return { code: 'nonfuture_job' };
      return { value: { job: op.job, due_time: op.due_time, status: 'pending' } as Json };
    case 'job.complete': {
      const ok =
        get(row, 'status') === 'pending' && (get(row, 'due_time') as number) <= ctx.horizon;
      return check(ok, put(row, { status: 'completed' }));
    }
    case 'time.advance':
      return check(row === op.from && op.to > op.from, op.to);
    case 'resource.adjust': {
      const spec = get(section(ctx.state, 'resource_specs'), key(op.resource)) as Spec | undefined;
      const ok =
        spec !== undefined &&
        current(row as Stored | undefined, spec, ctx.state.clock) === op.from &&
        op.to >= spec.minimum &&
        op.to <= spec.maximum;
      return check(ok, { value: op.to, at: ctx.state.clock });
    }
    case 'cooldown.start':
      return check(same(row, op.from) && op.at === ctx.state.clock, op.at);
  }
}

function choice(op: DeltaOp & { op: `choice.${string}` }, row: Json | undefined): Outcome {
  switch (op.op) {
    case 'choice.open': {
      const { actor_id, source, beat, roles, choice_ids } = op;
      const opened = { actor_id, source, beat, roles, choice_ids, status: 'pending' };
      return check(row === undefined, opened as unknown as Json);
    }
    case 'choice.resolve': {
      const offered =
        get(row, 'status') === 'pending' &&
        (get(row, 'choice_ids') as string[]).includes(op.choice_id);
      const ok = offered && get(row, 'opened_revision') === op.expected_revision;
      return check(ok, put(row, { status: 'resolved', choice_id: op.choice_id }));
    }
    default:
      return check(get(row, 'status') === 'pending', put(row, { status: 'closed' }));
  }
}

function quest(op: DeltaOp & { op: `quest.${string}` }, row: Json | undefined, ctx: Ctx): Outcome {
  if (op.op === 'quest.activate') {
    const taken = rows('quest', 'quests', 'instance_id', ctx).some(
      ([, r]) =>
        same(get(r, 'quest'), op.quest) &&
        same(get(r, 'scope'), op.scope) &&
        OPEN.includes(get(r, 'state') as string),
    );
    const created = { quest: op.quest, scope: op.scope, state: 'active' };
    return check(row === undefined && !taken, created as unknown as Json);
  }
  const outcomeOk =
    op.to === 'resolved'
      ? op.outcome !== undefined
      : op.to === 'failed' || op.outcome === undefined;
  const { outcome: _, ...rest } = (row ?? {}) as Obj;
  const next = op.outcome === undefined ? rest : { ...rest, outcome: op.outcome };
  const legal = (LEGAL[op.from] ?? []).includes(op.to);
  return check(get(row, 'state') === op.from && legal && outcomeOk, { ...next, state: op.to });
}

function transfer(e: string, source: string, d: string, row: Json | undefined, ctx: Ctx): Outcome {
  if (row !== source) return { code: 'precondition_failed' };
  if (inside(d, e, ctx)) return { code: 'containment_cycle' };
  const cap = get(section(ctx.state, 'capacities'), d) as number | undefined;
  if (cap === undefined) return { value: d };
  const held = rows('containment', 'containers', 'entity_id', ctx).filter(
    ([x, c]) => c === d && x !== e,
  );
  if (held.length >= cap) return { code: 'capacity_exceeded' };
  return { value: d };
}

// d is e or inside it. Revisiting a container means a cyclic base: fail closed.
function inside(d: string, e: string, ctx: Ctx): boolean {
  const seen = new Set<Json>();
  for (let at: Json | undefined = d; at !== undefined; at = read(containment(at as string), ctx)) {
    if (at === e || seen.has(at)) return true;
    seen.add(at);
  }
  return false;
}

// The overlay's value for a target, else the base's.
function read(t: MutationTarget, ctx: Ctx): Json | undefined {
  const w = ctx.overlay.get(key(t));
  if (w) return w.value;
  const s = ctx.state;
  switch (t.kind) {
    case 'fact':
      return get(section(s, 'facts'), key(t));
    case 'containment':
      return get(section(s, 'containers'), t.entity_id);
    case 'quest':
      return get(section(s, 'quests'), t.instance_id);
    case 'choice':
      return get(section(s, 'choices'), t.continuation_id);
    case 'job':
      return get(section(s, 'jobs'), t.job_id);
    case 'clock':
      return s.clock;
    case 'resource':
      return get(section(s, 'resources'), key(t));
    case 'cooldown':
      return get(section(s, 'cooldowns'), key(t));
  }
}

// ponytail: scans the whole section; add a contents/scope index when a cartridge has many rows.
function rows(kind: string, name: string, id: string, ctx: Ctx): [string, Json][] {
  const changed = new Map<string, Json>();
  for (const w of ctx.overlay.values())
    if (w.target.kind === kind) changed.set(get(w.target as Json, id) as string, w.value);
  const base = Object.entries(section(ctx.state, name)).filter(([k]) => !changed.has(k));
  return [...base, ...changed];
}
